dodiscover/context_builder.py

```diff
@@ -307,12 +311,206 @@ def _empty_graph_func(self, graph_variables) -> Callable:
         return empty_graph
 
 
-def make_context(context: Optional[Context] = None, create_using=ContextBuilder) -> ContextBuilder:
+class InterventionalContextBuilder(ContextBuilder):
+    """A builder class for creating observational+interventional data Context objects.
+
+    The InterventionalContextBuilder is meant solely to build Context objects that work
+    with observational + interventional datasets.
+
+    The context builder provides a way to capture assumptions, domain knowledge,
+    and data. This should NOT be instantiated directly. One should instead use
+    :func:`dodiscover.make_context` to build a Context data structure.
+
+    Notes
+    -----
+    The number of distributions and/or interventional targets must be set in order
+    to build the :class:`~.context.Context` object here.
+    """
+
+    _intervention_targets: Optional[List[Tuple[Column]]] = None
+    _num_distributions: Optional[int] = None
+    _obs_distribution: bool = True
+
+    def obs_distribution(self, has_obs_distrib: bool):
+        """Whether or not we have access to the observational distribution.
+
+        By default, this is True and assumed to be the first distribution.
+        """
+        self._obs_distribution = has_obs_distrib
+        return self
+
+    def num_distributions(self, num_distribs: int):
+        """Set the number of data distributions we are expected to have access to.
+
+        Note this must include observational too if observational is assumed present.
+        To assume that we do not have access to observational data, use the
+        :meth:`InterventionalContextBuilder.obs_distribution` to turn off that assumption.
+
+        Parameters
+        ----------
+        num_distribs : int
+            Number of distributions we will have access to. Will set the number of
+            distributions to be ``num_distribs + 1`` if ``_obs_distribution is True`` (default).
+        """
+        self._num_distributions = num_distribs
+        return self
+
+    def intervention_targets(self, targets: List[Tuple[Column]]):
```

Comment:
What about the values the interventions are set to?  I suppose this might not be needed for constraint based discovery, but perhaps it would be for SCM based discovery?

Reply:
There is no algorithm in dodiscover yet that learns from hard interventions. PsiFCI and IFCI learn from soft-interventions specifically. Eventually we will need to add that as more research is done on those methods. Whenever a new algorithm comes up, we can look into extending the context object.